lib/rex/io/socket_abstraction.rb

```diff
@@ -4,202 +4,216 @@
+              closed = true
+            end
 
-              # sf: Only remove the data off the queue is write was successfull.
-              #     This way we naturally perform a resend if a failure occured.
-              #     Catches an edge case with meterpreter TCP channels where remote send
-              #     failes gracefully and a resend is required.
-              if (sent.nil?)
+            unless closed
+              begin
+                buf = rsock.sysread(32_768)
+                if buf.nil?
+                  closed = true
+                  wlog('monitor_rsock: closed remote socket due to nil read')
+                end
+              rescue EOFError => e
                 closed = true
-                wlog("monitor_rsock: failed writing, socket must be dead")
-                break
-              elsif (sent > 0)
-                total_sent += sent
+                dlog('monitor_rsock: EOF in rsock')
+              rescue ::Exception => e
+                closed = true
+                wlog("monitor_rsock: exception during read: #{e.class} #{e}")
               end
-            rescue ::IOError, ::EOFError => e
-              closed = true
-              wlog("monitor_rsock: exception during write: #{e.class} #{e}")
-              break
             end
-          end
-        end
 
-        if( closed )
-          begin
-            self.close_write if self.respond_to?('close_write')
-          rescue IOError
+            unless closed
+              total_sent = 0
+              total_length = buf.length
+              while total_sent < total_length
+                begin
+                  data = buf[total_sent, buf.length]
+
+                  # Note that this must be write() NOT syswrite() or put() or anything like it.
+                  # Using syswrite() breaks SSL streams.
+                  sent = write(data)
+
+                  # sf: Only remove the data off the queue is write was successful.
+                  #     This way we naturally perform a resend if a failure occurred.
+                  #     Catches an edge case with meterpreter TCP channels where remote send
+                  #     fails gracefully and a resend is required.
+                  if sent.nil?
+                    closed = true
+                    wlog('monitor_rsock: failed writing, socket must be dead')
+                    break
+                  elsif sent > 0
+                    total_sent += sent
+                  end
+                rescue ::IOError, ::EOFError => e
+                  closed = true
+                  wlog("monitor_rsock: exception during write: #{e.class} #{e}")
+                  break
+                end
+              end
+            end
+
+            next unless closed
+
+            begin
+              close_write if respond_to?('close_write')
+            rescue StandardError
+            end
+
+            break
           end
-          break
+
+          rsock.sysclose
```

Comment:
This is now the one and only location where a monitored `rsock` will actually be closed. Other attempts to close it will result in the request being made and it's up to this loop to honor it and exit.